kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationPerUserSuite.scala

```diff
@@ -30,29 +33,118 @@ class KyuubiOperationPerUserSuite extends WithKyuubiServer with JDBCTests {
     KyuubiConf().set(KyuubiConf.ENGINE_SHARE_LEVEL, "user")
   }
 
-  test("ensure two connections in user mode share the same engine") {
-    var r1: String = null
-    var r2: String = null
-    new Thread {
-      override def run(): Unit = withJdbcStatement() { statement =>
-        val res = statement.executeQuery("set spark.app.name")
-        assert(res.next())
-        r1 = res.getString("value")
+  test("ensure two connections share the same engine when engine pool size is 1.") {
+    withSessionConf()(
+      Map(HighAvailabilityConf.HA_ZK_ENGINE_POOL_SIZE.key -> "1",
+        KyuubiConf.ENGINE_SHARE_LEVEL_SUB_DOMAIN.key -> "aaa"
+      ))(Map.empty) {
+
+      var r1: String = null
+      var r2: String = null
+      new Thread {
+        override def run(): Unit = withJdbcStatement() { statement =>
+          val res = statement.executeQuery("set spark.app.name")
+          assert(res.next())
+          r1 = res.getString("value")
+        }
+      }.start()
+
+      new Thread {
+        override def run(): Unit = withJdbcStatement() { statement =>
+          val res = statement.executeQuery("set spark.app.name")
+          assert(res.next())
+          r2 = res.getString("value")
+        }
+      }.start()
+
+      eventually(timeout(120.seconds), interval(100.milliseconds)) {
+        assert(r1 != null && r2 != null)
       }
-    }.start()
 
-    new Thread {
-      override def run(): Unit = withJdbcStatement() { statement =>
-        val res = statement.executeQuery("set spark.app.name")
-        assert(res.next())
-        r2 = res.getString("value")
+      assert(r1 === r2)
+    }
+  }
+
+  test("ensure two connections don't share the same engine when engine pool is 2.") {
+    withSessionConf()(
+      Map(HighAvailabilityConf.HA_ZK_ENGINE_POOL_SIZE.key -> "2",
+        KyuubiConf.ENGINE_SHARE_LEVEL_SUB_DOMAIN.key -> "bbb",
+        KyuubiConf.ENGINE_IDLE_TIMEOUT.key -> "30000"
+      ))(Map.empty) {
+
+      var r1: String = null
+      var r2: String = null
+      new Thread {
+        override def run(): Unit = withJdbcStatement() { statement =>
+          val res = statement.executeQuery("set spark.app.name")
+          assert(res.next())
+          r1 = res.getString("value")
+        }
+      }.start()
+
+      new Thread {
+        override def run(): Unit = withJdbcStatement() { statement =>
+          val res = statement.executeQuery("set spark.app.name")
+          assert(res.next())
+          r2 = res.getString("value")
+        }
+      }.start()
+
+      eventually(timeout(120.seconds), interval(100.milliseconds)) {
+        assert(r1 != null && r2 != null)
       }
-    }.start()
 
-    eventually(timeout(120.seconds), interval(100.milliseconds)) {
-      assert(r1 != null && r2 != null)
+      assert(r1 != r2)
     }
+  }
 
-    assert(r1 === r2)
+  test("ensure two of three connections share the same engine when engine pool size is 2.") {
```

Comment:
Can we add another case that connection 2, 3 concurrently created while connection 1's engine creation is finished

Reply:
Let me have a try.

Reply:
> Can we add another case that connection 2, 3 concurrently created while connection 1's engine creation is finished

hi @yaooqinn, what's the aim of the test case? what will we assert?

Reply:
we will exactly create 2 engines and one for 2 connections and the other for 1 connection

Reply:
Had updated the test case, correct me if wrong.